airflow/timetables/interval.py

```diff
@@ -211,6 +217,17 @@ def _align(self, current: DateTime) -> DateTime:
             return next_time
         return current
 
+    def _prev_align(self, current: DateTime) -> DateTime:
+        """Get the prev scheduled time.
+
+        This is ``current - interval``, unless ``current`` falls right on the
+        interval boundary, when ``current`` is returned.
+        """
+        prev_time = self._get_prev(current)
+        if self._get_next(prev_time) != current:
+            return prev_time
+        return current
```

Comment:
This seems to be always the same as `self._get_prev(self._get_next(current))`?

Say a schedule runs one every midnight. For this implementation:

* If `current` is at midnight, the return value is at midnight
* If `current` is on 1am, the return value is at midnight

For `self._get_prev(self._get_next(current))`:

* If `current` is at midnight, `_get_next` returns the next midnight, and `_get_prev` returns _this_ midnight
* If `current` is at 1am, `_get_next` returns the next midnight, and `_get_prev` returns _this_ midnight

And for the timedelta interval, `self._get_prev(self._get_next(current))` is always `current`.

So if I’m not mistaken, this function is not needed.

Reply:
They are different ha

For self._prev_align()
- If current is at 1am, the return value is at midnight 

For self._align()  
- If current is at 1am, the return value is next midnight.

It is correct that `data_schedule_start` is equal to the `data_schedule_end` of the previous one when no scheduling time changes. However, when the scheduling time changes, The first `data_schedule_start` for a new schedule may not be the same as the `data_schedule_end` before the schedule was changed
![2022-02-23_16-55](https://user-images.githubusercontent.com/55907021/155287695-aa251f60-df19-43e3-9d12-d8fb24be36c8.png)

Reply:
> This seems to be always the same as `self._get_prev(self._get_next(current))`?
> 
> Say a CRON schedule runs one every midnight. For this implementation:
> 
> * If `current` is at midnight, the return value is at midnight
> * If `current` is on 1am, the return value is at midnight
> 
> For `self._get_prev(self._get_next(current))`:
> 
> * If `current` is at midnight, `_get_next` returns the next midnight, and `_get_prev` returns _this_ midnight
> * If `current` is at 1am, `_get_next` returns the next midnight, and `_get_prev` returns _this_ midnight
> 
> And for the timedelta interval, `self._get_prev(self._get_next(current))` is always `current`.
> 
> So if I’m not mistaken, this function is not needed.

They are different ha

For self._prev_align()
- If current is at 1am, the return value is at midnight 

For self._align()  
- If current is at 1am, the return value is next midnight.

It is correct that `data_schedule_start` is equal to the `data_schedule_end` of the previous one when no scheduling time changes. However, when the scheduling time changes, The first `data_schedule_start` for a new schedule may not be the same as the `data_schedule_end` before the schedule was changed
![2022-02-23_16-55](https://user-images.githubusercontent.com/55907021/155287695-aa251f60-df19-43e3-9d12-d8fb24be36c8.png)

Reply:
Thanks for the breakdown. From this analysis it seems like a `_prev_align` is unavoidable. However, perhaps we could rename the functions to descibe things better. Maybe we could rename `_align` to `_align_to_next`, and name this new function `_align_to_prev`?

Reply:
Also we need to document pretty thoroughly _why_ we need both functions.